Consolidate notifications into one menu

benmanley opened the pull request:
Fixes #3123

@lefnire @paglias @deilann
I’ve styled up the notif dropdown, but the JS is stumping me as it’s not my forte. (I fiddled around with `User.user.newMessages’, but couldn’t quite achieve what I needed.) 

Grateful for any help or advice on implementing the following:

~~\- [ ] Hide the toolbar notif button if there are no new messages.~~
- [x] Add an inactive class to the toolbar notif button if there are no new messages.
- [x] Change the toolbar notif icon dependent on whether the newest message is an invitation or a message
- [x] Clear new messages individually using x button
- [ ] Clear all new messages using “Clear all” button

I also don’t know where and how to add new `=env.t’’` text (‘Notifs’, ‘Clear’, ‘Clear all’).
#### Desktop

![screen shot 2014-03-26 at 21 54 23](https://cloud.githubusercontent.com/assets/3317594/2531505/d0017c9e-b532-11e3-8322-c45be76d8192.png)
#### Mobile

![notifs-mobile](https://cloud.githubusercontent.com/assets/3317594/2531769/54f60368-b536-11e3-828c-51b75c6bde75.png)

In both views, notifs begin to scroll down if there are a lot of them.

(FYI, I’ve also tweaked the dropdown arrow next to Options.)

benmanley committed:
Consolidate notifications into one menu

benmanley committed:
Close notifs modal and toolbar when clicking an individual notif

paglias commented:
@benmanley maybe tomorrow I can code this but if you want to try:

you can show the notif button only when there's something to show (invitation or message) using:

`ng-if="user.invitations.party.id || user.invitations.guilds || user.newMessages"`

> Change the toolbar notif icon dependent on whether the newest message is an invitation or a message

I don't think we have a timestamp on the invitations... without it is difficult to achieve this

> Clear new messages individually using x button
> Clear all new messages using “Clear all” button

These two will be difficult too, we should add a property to newMessages, and invitations to set it as read or do you want to completely remove it from the user?

paglias commented:
I think messages can be removed without problems from user.newMessages but we can't delete invitations

benmanley commented:
Thanks @paglias. 

**Hide the toolbar notif button if there are no new messages.**
I tried `ng-if="user.invitations.party.id || user.invitations.guilds || user.newMessages"` and sent a new message.

`ng-if="user.invitations.party.id || user.invitations.guilds || user.newMessages"` didn't work
`ng-if="user.invitations.guilds || user.newMessages"` didn't work
`ng-if="user.invitations.party.id || user.newMessages"` worked
`ng-if="user.newMessages"` worked

**Change the toolbar notif icon dependent on whether the newest message is an invitation or a message**
Maybe date doesn't matter. Perhaps if we say that invitations are more important than messages, then would this work?
`span.glyphicon.glyphicon-user(ng-if='user.invitations.party.id || user.invitations.guilds ')`

Then I could style `span.glyphicon.glyphicon-user` to have a higher z-index than the notif icon.

The same goes for the mystery item. I guess that would have highest priority. I haven't added it to the notif system, but I could.

**Clear new messages individually using x button**
I think invitations could stay in the notif area until accepted, so, they wouldn't need to be deleted. Don't know if messages need to be deleted as I don't really understand what happens now when a notif is clicked and how the icon is dismissed.

lefnire commented:
*drools* This is looking awesome! Will see about the rest of this convo tomorrow, but I feel like you'll be a 7 in no time @benmanley. Glad to have you on the project :)

MaxHedrm commented:
Why not spell out "Notifications"? It seems silly to abbreviate it when there is plenty of room. :smile:

paglias commented:
@benmanley sorry, guilds is an object so we should check when it's empty... this should work:

`ng-if="user.invitations.party.id || user.invitations.guilds.length > 0 || _.isEmpty(user.newMessages)"`

> Change the toolbar notif icon dependent on whether the newest message is an invitation or a message

Oh yes, that would work I think, I didn't understood correctly before, you can use:

```
span.glyphicon.glyphicon-user(ng-if='user.invitations.party.id || user.invitations.guilds.length > 0 ')
span.glyphicon.glyphicon-another-icon(ng-if='user.newMessages && !user.invitations.party.id && user.invitations.guilds.length === 0 ')
```

But this way the user would not know if there are messages too...

> Clear new messages individually using x button

new messages are set as read using `Groups.seenMessage(group._id);` but it currently won't work from a view because Groups is not added to $rootScope, if @lefnire think it's fine we can add it

@MaxHedrm agree, we can use Notifications :)

@benmanley no problem about `env.t` tell me when it's merged so I can update it

benmanley commented:
Thanks @lefnire!

Thanks @paglias. I'll take a look. We have to decide whether it's more valuable to save toolbar space by combining invitations, messages, mystery items, etc than it is for the user to know if there are multiple new notification types. My guess is that people would tend to click invitations and mystery items straight away, so it wouldn't be a big problem, but I don't know if that's actually what most users do.    

@MaxHedrm
Yes. I used ‘Notifs’ because that’s what the HabitRPG guys seemed to call them, so I was following their convention. I think 'Notifications' is probably clearer too. I'll change it.

benmanley commented:
@lefnire @paglias 
I’ve been making some progress with the toolbar in general and will commit more shortly, but I'm still struggling to get the notification logic working.

``` jade
span.glyphicon.glyphicon-comment(ng-if='!user.invitations.party.id && !user.purchased.plan.mysteryItems.length || user.invitations.guilds.length === 0 && !user.purchased.plan.mysteryItems.length', ng-class='{inactive: _.isEmpty(user.newMessages)}')
span.glyphicon.glyphicon-user(ng-if='user.invitations.party.id && !user.purchased.plan.mysteryItems.length || user.invitations.guilds.length > 0 && !user.purchased.plan.mysteryItems.length')
span.glyphicon.glyphicon-gift(ng-if='user.purchased.plan.mysteryItems.length')
```

What I'm trying to do is:

If there’s a new mystery item, show glyphicon-gift
Else if there’s a party or guild invitation, show glyphicon-user
Else show glyphicon-comment and, if there are no messages, apply the class “inactive”.

The following don’t seem to work:
`_.isEmpty(user.newMessages)`
`user.invitations.guilds.length === 0`

Any ideas?

paglias commented:
@benmanley I would use a single `span` element and apply the relative classes with `ng-class` (http://docs.angularjs.org/api/ng/directive/ngClass for examples).

About the code

```
span.glyphicon.glyphicon-comment(ng-if='!user.invitations.party.id && !user.purchased.plan.mysteryItems.length || user.invitations.guilds.length === 0 && !user.purchased.plan.mysteryItems.length', ng-class='{inactive: _.isEmpty(user.newMessages)}')
```

why not:

```
span.glyphicon.glyphicon-comment(ng-if='!user.invitations.party.id && !user.purchased.plan.mysteryItems.length && user.invitations.guilds.length === 0', ng-class='{inactive: _.isEmpty(user.newMessages)}')
```

Then:

```
span.glyphicon.glyphicon-user(ng-if='!user.purchased.plan.mysteryItems.length && (user.invitations.party.id ||  user.invitations.guilds.length > 0)')
```

The third one is ok, but I would use ngClass, otherwise you risk that more than a class is shown when there are both messages and inviatations or messages and mystery items.

About `_.isEmpty(user.newMessages)` and `user.invitations.guilds.length === 0` would you mind doing some testing for me that I can't actually?

Open /public/js/controllers/authCtrl.js

and add these two watchers in the controller:

```
// Logs user.newMessages and result of _.isEmpty(user.newMessages)
$scope.$watch('user.newMessages', function(value, old){console.log(value, _.isEmpty(value))});

// Logs user.invitations.guilds and it's length
$scope.$watch('user.invitations.guilds', function(value, old){console.log(value, value.length)});
```

a look at the output in the console (if it doesn't work change $scope.$watch with $rootScope.$watch), and report it back here

benmanley commented:
Thanks @paglias. I'll give this a try and report back.

benmanley commented:
@paglias Here's the output. Let me know if that's what you were looking for. (If it's relevant, there were no messages in the notification window when I logged it.)

``` javascript
[Log] Object (authCtrl.js, line 93)
1aac143b-1cc8-48b5-b46f-53c54b05b4bc: Object
name: "Guild One"
value: false
__proto__: Object
4ec65dd7-5751-410e-ac2f-15578efbb7fa: Object
name: "Test Guild"
value: false
__proto__: Object
30be90b0-b78b-4f2c-8d39-bfd76ebdd397: Object
value: false
__proto__: Object
60979cf5-5feb-4810-82fa-212d0c239d64: Object
name: "Guild Two – This one has a really long name so everyone can see what happens if someone goes crazy naming their guild. If you thought "Sedulous Sesquipedalians" was seriously long, think again!"
value: false
__proto__: Object
cfd1351a-cdb6-4153-a944-edd1e297b0bd: Object
name: "Test Party"
value: false
__proto__: Object
undefined: false
__proto__: Object
__defineGetter__: function __defineGetter__() {
__defineSetter__: function __defineSetter__() {
__lookupGetter__: function __lookupGetter__() {
__lookupSetter__: function __lookupSetter__() {
constructor: function Object() {
get __proto__: function () {
hasOwnProperty: function hasOwnProperty() {
isPrototypeOf: function isPrototypeOf() {
propertyIsEnumerable: function propertyIsEnumerable() {
set __proto__: function () {
toLocaleString: function toLocaleString() {
toString: function toString() {
valueOf: function valueOf() {
 false
[Log] [] 0 (authCtrl.js, line 96)
[Log] Object (authCtrl.js, line 93)
1aac143b-1cc8-48b5-b46f-53c54b05b4bc: Object
name: "Guild One"
value: false
__proto__: Object
4ec65dd7-5751-410e-ac2f-15578efbb7fa: Object
name: "Test Guild"
value: false
__proto__: Object
30be90b0-b78b-4f2c-8d39-bfd76ebdd397: Object
value: false
__proto__: Object
60979cf5-5feb-4810-82fa-212d0c239d64: Object
name: "Guild Two – This one has a really long name so everyone can see what happens if someone goes crazy naming their guild. If you thought "Sedulous Sesquipedalians" was seriously long, think again!"
value: false
__proto__: Object
cfd1351a-cdb6-4153-a944-edd1e297b0bd: Object
name: "Test Party"
value: false
__proto__: Object
undefined: Object
value: false
__proto__: Object
__proto__: Object
__defineGetter__: function __defineGetter__() {
__defineSetter__: function __defineSetter__() {
__lookupGetter__: function __lookupGetter__() {
__lookupSetter__: function __lookupSetter__() {
constructor: function Object() {
get __proto__: function () {
hasOwnProperty: function hasOwnProperty() {
isPrototypeOf: function isPrototypeOf() {
propertyIsEnumerable: function propertyIsEnumerable() {
set __proto__: function () {
toLocaleString: function toLocaleString() {
toString: function toString() {
valueOf: function valueOf() {
 false
[Log] [] 0 (authCtrl.js, line 96)
```

paglias commented:
@benmanley strange.. evertything seems ok, guild invitations is an empty array when there's nothing so length should work, same for newMessages. Will test on my own tomorrow but you'll have to wait a bit, sorry

benmanley commented:
No problem. Thanks.

paglias commented:
@benmanley ok, `.isEmpty` is correct, there's an error in the server side code:

https://github.com/HabitRPG/habitrpg/blob/develop/src/controllers/groups.js#L257-L260 should be changed to, but would work with new users only, we need to run a migration , in the meantime assume it's working (or change the code with the one below and register a new user to have it working, I'm not changing it immediatelly because it would work only for new signups so I prefer to do everything together, but it's working)

```
  var update = {$set:{}},
      gid = req.params.gid;
  console.log(gid)
  if(gid){
    update['$set']['newMessages.'+req.params.gid+'.value'] = false;
    User.update({_id:req.headers['x-api-user'], apiToken:req.headers['x-api-key']},update).exec();
  }
```

---

`user.invitations.guilds.length === 0` is working, I don't know if you tried with the code I posted yesterday but try using:

```
        span.glyphicon.glyphicon-comment(ng-if='!user.invitations.party.id && !user.purchased.plan.mysteryItems.length && user.invitations.guilds.length === 0', ng-class='{inactive: _.isEmpty(user.newMessages)}')
        span.glyphicon.glyphicon-user(ng-if='!user.purchased.plan.mysteryItems.length && (user.invitations.party.id ||  user.invitations.guilds.length > 0)')
        span.glyphicon.glyphicon-gift(ng-if='user.purchased.plan.mysteryItems.length')
```

Which has some more parenthesis than your code, and let me know if it works

benmanley committed:
Start a custom CSS framework

benmanley committed:
Redesign toolbar navigation and fix notification logic

benmanley committed:
Remove top level tabbed navigation and gem wallet

benmanley commented:
@paglias Yes. I tried your code and I think it’s working now. (Haven’t got it working with ng-class yet, though). Thanks!

Here’s what else I’ve been up to:
- Separated toolbar submenu into individual drop downs
- Combined Settings and Help and moved to right side of toolbar
- Removed the top level nav tabs from the area below the header
- Started an HRPG custom CSS framework with global styles for colors, buttons and modals.

I’ve committed my changes so far, but it’s not quite ready to go. I need to do a bit more work on the mobile toolbar and some of the fine detail.

Have you had any more thoughts on clearing notifications (individually and in bulk)? I’ve commented out the buttons for now.

![screen shot 2014-04-03 at 00 32 00](https://cloud.githubusercontent.com/assets/3317594/2600967/2e83dd62-bb09-11e3-85ff-edd84e20bbc1.png)

deilann commented:
Whilst you do this, is there any way we can also once again have the
ability to collapse the toolbar if we want to?

On Thu, Apr 3, 2014 at 1:24 AM, Ben Manley notifications@github.com wrote:

> @paglias https://github.com/paglias Yes. I tried your code and I think
> it's working now. (Haven't got it working with ng-class yet, though).
> Thanks!
> 
> Here's what else I've been up to:
> - Separated toolbar submenu into individual drop downs
> - Combined Settings and Help and moved to right side of toolbar
> - Removed the top level nav tabs from the area below the header
> - Started an HRPG custom CSS framework with global styles for colors,
>   buttons and modals.
> 
> I've committed my changes so far, but it's not quite ready to go. I need
> to do a bit more work on the mobile toolbar and some of the fine detail.
> 
> Have you had any more thoughts on clearing notifications (individually and
> in bulk)? I've commented out the buttons for now.
> 
> [image: screen shot 2014-04-03 at 00 32 00]https://cloud.githubusercontent.com/assets/3317594/2600967/2e83dd62-bb09-11e3-85ff-edd84e20bbc1.png
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/pull/3155#issuecomment-39424866
> .

paglias commented:
@benmanley  that's fabulous! Clearing notifications (for messages not invitations) is possible, just call Groups.seenMessage(groupId) for every group's messages you want to set as seen. If you add the html and css I can I can add the JS for you

benmanley committed:
Improve toolbar and global styles

benmanley commented:
@paglias Not sure if I’ve got this right, but here’s the code for clearing messages ...

``` jade
li(ng-repeat='(k,v) in user.newMessages', ng-if='v.value')
            a(ng-click='k==party._id ? $state.go("options.social.party") : $state.go("options.social.guilds.detail",{gid:k}); user.preferences.isNotifsExpanded = !user.preferences.isNotifsExpanded; set({"preferences.menuExpanded":false})')
              span.glyphicon.glyphicon-comment
              span {{v.name}}
            a(ng-click='Groups.seenMessage(k)')
              span.glyphicon.glyphicon-remove-circle(popover=env.t('clear'),popover-placement='right',popover-trigger='mouseenter',popover-append-to-body='true')
```

I’ve also:
- Added a message when the notifications dropdown is empty.
- Converted px values to em where appropriate
- Narrowed the toolbar a bit
- Colour-coded the currency text (see #3076)
- Improved the mobile toolbar
- Added a mixin to allow easy customisation of button colours

Still haven’t quite nailed the mobile view. This might help...

**Rounding down GP to the nearest thousand**

I wanted to round down GP (and maybe gems) over 999 to the format: 1.0K

Something like:

``` js
if (gp > 999) {
  return Math.floor(gp / 100) / 10.0 + “K”;
}
else {
  return Math.floor(gp);
}
```

Not sure where to put the code though!

paglias commented:
@benmanley the code is ok, but have you tried it? I'm not sure Groups is avalaible from views. If it doesn't work (and it shouldn't) go to [rootCtrl.js](https://github.com/HabitRPG/habitrpg/blob/develop/public/js/controllers/rootCtrl.js) and add 

`$rootScope.Groups = Groups;` under this line https://github.com/HabitRPG/habitrpg/blob/develop/public/js/controllers/rootCtrl.js#L29

I would move the popover to the `<a>` tag:

``` jade
            a(ng-click='Groups.seenMessage(k)', popover=env.t('clear'),popover-placement='right',popover-trigger='mouseenter',popover-append-to-body='true')
              span.glyphicon.glyphicon-remove-circle
```

**Rounding down GP to the nearest thousand**

The number is already floored https://github.com/HabitRPG/habitrpg-shared/blob/develop/script/index.coffee#L218, you should use a filter, edit https://github.com/HabitRPG/habitrpg/blob/develop/public/js/filters/filters.js to this:

``` js
angular.module('habitrpg')
  .filter('gold', function () {
    return function (gp) {
      return Math.floor(gp);
    }
  })
  .filter('silver', function () {
    return function (gp) {
      return Math.floor((gp - Math.floor(gp))*100);
    }
  })
  .filter('htmlDecode',function(){
    return function(html){
      return $('<div/>').html(html).text();
    }
  })
  .filter('goldThousands', function(){
    return function (gp) {
      return (gp > 999) ? (Math.floor(gp / 100) / 10.0 + “K”) : gp;
    }
  })
```

Feel free to change the filter name, I used the first one that come to my mind.

And then use it [here](https://github.com/HabitRPG/habitrpg/blob/develop/views/shared/header/menu.jade#L40):

`| {{Shared.gold(user.stats.gp)}} | goldThousands`

benmanley commented:
Grazie @paglias. Yes, clearing messages worked only when I added `$rootScope.Groups = Groups;`.

I think something’s wrong with `| {{Shared.gold(user.stats.gp)}} | goldThousands`. It doesn’t seem to be closed correctly as the filter appears as text, e.g.

![screen shot 2014-04-05 at 19 21 43](https://cloud.githubusercontent.com/assets/3317594/2623450/aa59ce5a-bcef-11e3-99ca-c38e259faa6e.png)

paglias commented:
@benmanley ops, the code filter should be between the brackets, sorry:

| {{Shared.gold(user.stats.gp) | goldThousands}}

benmanley commented:
Non si preoccupi. That's what I wondered, but `| {{Shared.gold(user.stats.gp) | goldThousands}}` made things go seriously freaky. I'm talking floating snowman freaky.

![screen shot 2014-04-05 at 19 47 49](https://cloud.githubusercontent.com/assets/3317594/2623492/e68b989c-bcf2-11e3-989f-900cb2861463.png)

Turns out the filter didn’t like the double quotes around K.

I changed `10.0 + “K”` to `10.0 + ‘K’` and everything is perfect. Grazie mille. (My Italian is about as good as my Javascript :wink: )

paglias commented:
'Non si preoccupi.' is just a little formal but grammatically perfect :)
Il 05/apr/2014 21:12 "Ben Manley" notifications@github.com ha scritto:

> Non si preoccupi. That's what I wondered, but | {{Shared.gold(
> user.stats.gp) | goldThousands}} made things go seriously freaky. I'm
> talking floating snowman freaky.
> 
> [image: screen shot 2014-04-05 at 19 47 49]https://cloud.githubusercontent.com/assets/3317594/2623492/e68b989c-bcf2-11e3-989f-900cb2861463.png
> 
> Turns out the filter didn't like the double quotes around K.
> 
> I changed 10.0 + "K" to 10.0 + 'K' and everything is perfect. Grazie
> mille. (My Italian is about as good as my Javascript [image: :wink:] )
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/pull/3155#issuecomment-39647971
> .

benmanley commented:
There are two reasons for that: 1. I wanted to be polite. 2. I learn all my Italian from [Inspector Montalbano](http://www.memrise.com/course/73147/inspector-montalbano-italian/) ;-)

benmanley commented:
@deilann re: collapsible toolbar – Now that top level navigation and currency are only in the toolbar, it seems right for the toolbar to be sticky, at least for desktop. 

I suppose the main alternatives would be:
1. Reinstate the toolbar toggle button
2. Toggle toolbar stickiness via settings
3. Do a Pinterest-style disappearing toolbar that reappears when you scroll back up the page.

Personally, I would rather avoid adding another button to the toolbar, but I think this might be one of those situations that’s worth a poll. My preference would probably be for 2, if anything, but maybe it needs debating further.

deilann commented:
This isn't a stickiness issue -- this is a people want to hide some things
and not see them for productivity reasons.

benmanley commented:
I see. So are we talking about Bailey messages and notifications in particular? Perhaps there could be a "Mute" button in notifications that hides/unhides messages and Bailey.

deilann commented:
Also things like the gems and the coins. Some people aren't worried about
distractions, but want to look more professional at work. If you turn off
the header and hide the toolbar, it can look very much like any other
productivity app.

On Sun, Apr 6, 2014 at 3:24 AM, Ben Manley notifications@github.com wrote:

> I see. So are we talking about Bailey messages and notifications in
> particular? Perhaps there could be a "Mute" button in notifications that
> hides/unhides messages and Bailey.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/pull/3155#issuecomment-39664015
> .

benmanley committed:
Add filter to round gold > 999 to nearest thousand

benmanley committed:
Add Groups to $rootScope

Allows messages to be cleared from a view

benmanley committed:
Replace text with env.t and remove unnecessary code

benmanley committed:
Improve toolbar styles

benmanley commented:
Ciao @paglias. I’ve committed some more changes. I think it's done. There’s just one more thing I need your help with: 

**Close open dropdown when another is clicked**
What’s the best way to close an open dropdown when another dropdown icon is clicked, so that only one dropdown can be open at a time?

Also, there are some things I think you’ll need to do once the changes are merged:

**env.t**
- menu
- notifications
- clear
- gold – Can we display the full amount of gold in the popover using `env.t(‘gold’)`?

**.isEmpty**

> https://github.com/HabitRPG/habitrpg/blob/develop/src/controllers/groups.js#L257-L260 should be changed to, but would work with new users only, we need to run a migration.

And a couple of optional things we could add:

**Gems filter**
I think it’s unlikely that anyone could rack up more than 999 gems, but we might want to put a filter on just in case.

**Clear all**
I don’t know if it’s possible to clear all new messages at once. 

Here are some screenshots:

**Desktop**

![screen shot 2014-04-08 at 16 40 08](https://cloud.githubusercontent.com/assets/3317594/2645702/87fdbc20-bf35-11e3-96cd-3a347d52daa1.png)

**Mobile**

![screen shot 2014-04-08 at 16 40 40](https://cloud.githubusercontent.com/assets/3317594/2645704/88021874-bf35-11e3-8252-76e6bfd2ebb9.png)

![screen shot 2014-04-08 at 16 41 20](https://cloud.githubusercontent.com/assets/3317594/2645700/87fb5534-bf35-11e3-899c-9a3767cc2e3b.png)

![screen shot 2014-04-08 at 16 42 07](https://cloud.githubusercontent.com/assets/3317594/2645701/87fc919c-bf35-11e3-86b9-7242d6072df2.png)

![screen shot 2014-04-08 at 16 41 06](https://cloud.githubusercontent.com/assets/3317594/2645703/87fe43e8-bf35-11e3-915a-4762901a92cc.png)

paglias commented:
Hi @benmanley ! Would you mind adding me as a collaborator to your fork so that I can work directly there?

paglias commented:
Regarding @benmanley 's last comment, everythign you said should be possible to do

benmanley commented:
@paglias Va bene. I've added you.

Mihakuu commented:
would the mobile always be doubled lined? or would it shorten down to take less screen space?

benmanley commented:
@Mihakuu If a user subscribes, that button will disappear. So, once we consolidate the currency icons, they should also fit on the top line for subscribers.

I don’t know yet whether Bailey notifications will fit too. If not, they might need to be incorporated into the notifications menu for mobile or maybe pop up from the bottom of the screen.

paglias commented:
( @benmanley having a busy time at school and other things, I hope to have some more free time next week )

benmanley commented:
Non c'è problema.

Mihakuu commented:
@benmanley  I know that we'd love for them to sub, but I fear a backlash of people being upset or angry at being 'forced' to sub to get rid of that second line of tool bar (unless you do add back in the toggle tool bar like @deilann suggested).

 Just looking at my phone, seeing how much is covered by the tool bar now and adding enough room for a Bailey sprite would take up a huge portion of my screen to the point I wouldn't use the mobile web version because so little would fit in the space left.

benmanley commented:
@Mihakuu I remember seeing somewhere that @lefnire wanted the subscription button to be fairly prominent. But I agree that it’s rather intrusive at the moment. 

A more subtle option might be to add the subscription button to the mobile submenu screen. Maybe next to the close button.

Regarding toggling, there seem to be three main reasons for wanting to toggle. 

**Distraction**
I think distraction is better solved by ‘muting’ notifications.

**Looking professional**
Looking more professional is probably better achieved by having a non-RPG mode which disables the header and the sprite sheet with fallbacks for certain images.

**Increasing screen space**
One alternative is to switch off toolbar stickiness via settings. 
@deilann suggested toggling via the settings dropdown, but I’m not entirely sold on that because I think it’s slightly cumbersome to activate and I imagine you'll be toggling all the time in order to navigate. 
Yet another alternative would be to activate a toggle button via settings for those who want it.

We might need a poll to work out what the demand is for all these features. Then we could decide which to implement.

Mihakuu commented:
I really don't mind leaving the sub button as the prominate thing on the bar before subbing. I just saying that having the bar being so large two lines. Maybe if the second line wasn't sticky or something? I just know that if we cover up a third of a users screen with a tool bar they're not going to like it.

deilann commented:
I don't understand why it's such a big deal to be able to have the toolbar
collapsable.

On Fri, Apr 11, 2014 at 2:22 PM, Mihakuu notifications@github.com wrote:

> I really don't mind leaving the sub button as the prominate thing on the
> bar before subbing. I just saying that having the bar being so large two
> lines. Maybe if the second line wasn't sticky or something? I just know
> that if we cover up a third of a users screen with a tool bar they're not
> going to like it.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/pull/3155#issuecomment-40254676
> .

lefnire commented:
Sorry for the MIA, I'm pushing this stuff to beta right now so people can start playing around. really liking the changes @benmanley!

lefnire commented:
@benmanley @paglias how do we feel about this PR's readiness? I've played with it locally & am happy. I added the missing strings to habitrpg-shared. Ben, once we get this merged - I've added you as a repo admin so you can make direct commits to tidy up any of your own work (or hotfixes in other locations too). Do continue to submit PRs for anything big / potentially-controversial.

paglias commented:
@benmanley @lefnire I gave it a look and think it's a big improvement over the current situation. But before using on prod there are a few things that I would like to see addressed:
1. The biggest one: close dropdown after a click happens anywhere in the page ([ngBlur](http://docs.angularjs.org/api/ng/directive/ngBlur) should do the trick
2. "Tasks" button should be "Options" (or another name) when in the tasks view and Tasks when in another view - MINOR
3. In the dropdown menus the links are clickable only over the text, the area around it should be clickable too - I would remove the padding from the <ul> and the margin from the <li> and add them to the <a>

I'll see if I can do anything about these later

benmanley commented:
@paglias Agree with 1 & 3. Not sure about 2. If the text for “Tasks” and “Options” is even a little different in length, the other buttons will jump left or right when you click between the two views. Could be distracting.

We might also need to fix #3241. Don’t know how it was done previously.

lefnire commented:
Same here: I agree with 1, not so much with 2. As for 1, it might be easier to code clicking another drop-down will close other open dropdowns (or just the previously-clicked one)? My thinking is ngBlur might be hard to code such that it doesn't close while you're trying to mouse over to one of the items in the drop-down. Play with it, whatever's easiest to code IMO.

Also, you're gonna kill me (and feel free to ignore this suggestion) - but now that it's popovers instead of header-replacment, `.glyphicon-chevron-down` might make more sense than menu-lines? *hides*

wc8 commented:
Hi, I just noticed the Help menu is missing... I was going to suggest a link to the Newbies guild might be useful there. Is that coming back or getting scratched?

As for 2, if the task button needs a function, perhaps it eventually could drop down with a few favorite tag filter shortcuts or, maybe even an Show All tasks item. Regardless, like having it there. 
If more room is needed, perhaps an icon :white_check_mark: could replace the word?

paglias commented:
@benmanley @lefnire ok I'll go with 1,3 and try fixing #3241 but I have a problem, a stupid one... this  PR isn't merged yet so where is the site being deployed on beta? from @benmanley 's benmaley/habitrpg?

lefnire commented:
@paglias i'm doing some multi-branch merging magic on local to push to beta. Heroku lets you deploy arbitrary SHAs from local, which rocks. I haven't pushed in a hot minute though, let me do that now

benmanley commented:
@deilann
Not a big deal to add it, but I’m looking from the opposite way round. Why do we need it at all? As top level navigation and currency are only in the toolbar now, I imagine users would be toggling the toolbar so frequently that it would be unnecessary. 

You may be right, but I feel like the best way to judge that is to let the new layout settle for a while, then do a poll to ask users what they think is missing/malfunctioning.

@wc8 I’ve included the help links in Settings. I’ve noticed sites like Google, Facebook and Twitter doing this and thought it would save a bit of space in the toolbar. But, if people can't find it, we may need to add an extra icon.

@lefnire 

> `.glyphicon-chevron-down` might make more sense than menu-lines?

:flips table: 

I’ll ponder chevrons vs menu icons. I’ve been continuing on this PR working to tie in the navigation further down the page. Shall I commit this now or hold off?

![subnav](https://cloud.githubusercontent.com/assets/3317594/2713498/c0e92940-c4e7-11e3-9d21-6e12ed7b52c9.png)

![screen shot 2014-04-15 at 22 42 16](https://cloud.githubusercontent.com/assets/3317594/2713497/c0e8c770-c4e7-11e3-88e5-6a3a6962e636.png)

lefnire commented:
@benmanley hahah, Just kidding, ignore me! Looks good :) Latest stuff on beta btw

deilann commented:
Many users have already made it clear though that they want the
functionality back. It was a feature that used to exist and then was taken
away for no real reason. I highly doubt users will be toggling and
untoggling all the time. If they want to be navigating, I would guess that
they would have the toolbar out. If they feel like the way they need to use
the site at that moment isn't toolbar dependant and they wish not to have
it, they'll hide it until they need it. I understand you feel like it's
unnecessary, but I and many others would like that functionality.
Considering it used to exist and does not hinder other users who do not
wish to use it, I just cannot understand why you are so opposed. I am not
trying to bully you here. I just really don't understand and it seems like
from this it is because you don't want it AMD are assuming others don't as
well
On Apr 15, 2014 2:57 PM, "Ben Manley" notifications@github.com wrote:

> @deilann https://github.com/deilann
> Not a big deal to add it, but I’m looking from the opposite way round. Why
> do we need it at all? As top level navigation and currency are only in the
> toolbar now, I imagine users would be toggling the toolbar so frequently
> that it would be unnecessary.
> 
> You may be right, but I feel like the best way to judge that is to let the
> new layout settle for a while, then do a poll to ask users what they think
> is missing/malfunctioning.
> 
> @wc8 https://github.com/wc8 I’ve included the help links in Settings.
> I’ve noticed sites like Google, Facebook and Twitter doing this and thought
> it would save a bit of space in the toolbar. But, if people can't find it,
> we may need to add an extra icon.
> 
> @lefnire https://github.com/lefnire
> 
> .glyphicon-chevron-down might make more sense than menu-lines?
> 
> :flips table:
> 
> I’ll ponder chevrons vs menu icons. I’ve been continuing on this PR
> working to tie in the navigation further down the page. Shall I commit this
> now or hold off?
> 
> [image: subnav]https://cloud.githubusercontent.com/assets/3317594/2713498/c0e92940-c4e7-11e3-9d21-6e12ed7b52c9.png
> 
> [image: screen shot 2014-04-15 at 22 42 16]https://cloud.githubusercontent.com/assets/3317594/2713497/c0e8c770-c4e7-11e3-88e5-6a3a6962e636.png
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/pull/3155#issuecomment-40539620
> .

Mihakuu commented:
Is it hard to add in the collapsible toolbar? Especially on mobile, I'd prefer if off to keeping such a large portion of my screen covered with it...

benmanley committed:
Add global styles for labels and menus

benmanley committed:
Replace bootstrap classes in submenus

benmanley committed:
Add custom styles for submenus

benmanley committed:
Restyle avatar popover

benmanley commented:
@deilann Please don’t think I took it away for no reason. I’m trying to make the toolbar clearer and remove anything extraneous. My gut feeling was that it was not a feature that the majority of people would want to use after the changes had been made. It's been said before that we can't please all of the people all of the time, so I stuck my neck out and removed it.

If it’s a genuinely useful feature to a large number of users, I’ll gladly reinstate it. I suggested a poll to determine whether that's the case, rather than relying solely on my (or your) opinion.

deilann commented:
Maybe you missed my suggestion-- I pointed out that it could easily go in
the settings drop down so it wouldn't take any more space. I could see
doing a poll, but the insinuation that it's not desired or wanted is very
hurtful when people have repeatedly spoken up wanting it. Desired by the
majority or a large number of users is a different beast but i still
disagree with that a poll is needed at all as it would benefit some users
and not inconvenience those who don't want it. It's also a feature that was
taken away, not a feature to be added and I recall users being annoyed by
that in the Tavern. Once something exists, taking it away requires, in my
opinion, more reason than adding in something new.
On Apr 15, 2014 4:13 PM, "Ben Manley" notifications@github.com wrote:

> @deilann https://github.com/deilann Please don’t think I took it away
> for no reason. I’m trying to make the toolbar clearer and remove anything
> extraneous. My gut feeling was that it was not a feature that the majority
> of people would want to use after the changes had been made. It's been said
> before that we can't please all of the people all of the time, so I stuck
> my neck out and removed it.
> 
> If it’s a genuinely useful feature to a large number of users, I’ll gladly
> reinstate it. I suggested a poll to determine whether that's the case,
> rather than relying solely on my (or your) opinion.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/pull/3155#issuecomment-40545870
> .

lefnire commented:
Ok y'all, got some hefty stuff I need to dev against `develop`. So we're back to production-tracks-master, beta-tracks-develop. Snapped master at 56ab738 , merging this one into develop. Keep the convo going, don't mind the closed status. @benmanley feel free to direct-commit to HabitRPG/develop

benmanley commented:
@deilann Don’t worry, I’m not insinuating anything. I thought a survey would give us some numbers to work with, but I know you interact with tons of users all the time, so I will bow to your better judgement.

I did see your suggestion about the settings icon (I’ve commented on it further up this thread somewhere). I reckon reinstating a button of some kind is a better option and I guess you'd probably prefer that anyway. I slept on it and a few ideas came to me in a dream. I’ll work on it next.

deilann commented:
You are _awesome_ Ben. I honestly doesn't care where it is, just as long as
the functionality is available without going to a settings page. So just
_somewhere_ accessible from the bar.

On Wed, Apr 16, 2014 at 1:43 PM, Ben Manley notifications@github.comwrote:

> @deilann https://github.com/deilann Don’t worry, I’m not insinuating
> anything. I thought a survey would give us some numbers to work with, but I
> know you interact with tons of users all the time, so I will bow to your
> better judgement.
> 
> I did see your suggestion about the settings icon (I’ve commented on it
> further up this thread somewhere). I reckon reinstating a button of some
> kind is a better option and I guess you'd probably prefer that anyway. I
> slept on it and a few ideas came to me in a dream. I’ll work on it next.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/pull/3155#issuecomment-40649723
> .

benmanley commented:
Got it. Thanks @deilann.

lefnire commented:
@benmanley see https://github.com/HabitRPG/habitrpg/commit/259aa865d52ccf8bd642d6d480e421b47d546add . I haven't merged it yet, part of a dev branch I'm on. It fixes the "only one menu item open at a time". Going through the code, I may have missed some conversation - I don't think we want to persist menu-expansion to the user in the database. I can see what you're getting at with it, in case they refresh they get the same open menu as when they left it - but I think it's a pretty dirty solution. You ok with this, having menu-expansion a local variable?

ghost commented:
@benmanley @paglias @lefnire Sorry for getting into serious guys' discussion, but let me thank you for implementing the feature so much wanted before! And a quick question: is it possible to have notifications about new challenges (show the challenge icon before the guild name in notifications list) and new guilds (the name of the guild with the 'new' icon)?

deilann commented:
Oh, those would be nice...

On Thu, Apr 17, 2014 at 12:05 AM, Sphinx notifications@github.com wrote:

> @benmanley https://github.com/benmanley @pagliashttps://github.com/paglias
> @lefnire https://github.com/lefnire Sorry for getting into serious
> guys' discussion, but let me thank you for implementing the feature so much
> wanted wanted before! And a quick question: is it possible to have
> notifications about new challenges (show the challenge icon before the
> guild name in notifications list) and new guilds (the name of the guild
> with the 'new' icon)?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/pull/3155#issuecomment-40687001
> .

paglias commented:
@benmanley I've enlarged the clickable area https://github.com/HabitRPG/habitrpg/commit/2eea5c2dbda00b06eed1211d21f2af2934263822 (only when not collapsed -> https://github.com/HabitRPG/habitrpg/commit/81ef26087f17202aaf192aff0f5223705ee7a2e8)

I've used two `!important` because I had difficulties changing some styles without changing behavior in other parts. If you can find a better solution, please use it

paglias commented:
@benmanley I've been a little absent these days, if you need anything or there is anythign that I said I would have done and I didn't can you remember me?

ArlieS commented:
The get-my-messages icon in the toolbar badly needs a tooltip. (The one to the right of the silver piece count) Until I found this thread, I figured you'd removed notifications entirely, The icon on the far right - just before the "subscribe" button (looks like a gear wheel) also lacks a tool tip. 

Also, as a general comment - when the UI changes drastically, some users would very much like to get an explanatory notification. This is particularly true in that many of us will see the change when we first get out of bed and are not awake.  I found this thread by going to the Back Corner basically to bitch about the clearly degraded UI, whereupon Bailey gave me this link. The image early in the thread told me what the thing-that-might-be-a-cartoon-style-talk-bubble was intended to do. (Still haven't checked whether it works ;-))

Mihakuu commented:
Can I suggest that the chat bubble change color when there is at least one notification available? At this point it's forcing you to click on it to check.

MaxHedrm commented:
Perhaps an outline (empty bubble) when there are no messages & the current filled in one when there are messages.

ArlieS commented:
AND an explanatory tooltip, as users may well not even notice the difference

wc8 commented:
See #3270 re: notification visual alert.

@lefnire / @lemoness  it does seem like it'd be a good idea to have the Town Crier announce when the interface gets an update. Ideally with a little overview: e.g. _the much anticipated consolidation of notifications has arrived_ or _as requested, your notifications are now grouped together in the notifications menu..._

Mihakuu commented:
I don't understand why the mobile tool bar couldn't have stayed compact like it was. Even subbed it takes two lines for everything to fit. It's very large

wc8 commented:
_Updated_: I wouldn't want the two lines either... In my case the second line is entirely blank, which makes it especially puzzling. 

What if we were to compact it like so: 
- [ ] **Menu**, Subscribe, **Gems**, **Gold**, **(Crier)**

Hopefully that all fits on a line, but if it doesn't, well, subscribing should make it fit on one line.
Menu would cover everything else from Tasks to Settings. Possibly exact coin could be shown via drop down.

P.S. If the added height is because of the town crier, what if we just use an icon on the mobile app or... can bailey peek more? e.g. only the eyes and top of head showing?

benmanley commented:
@lefnire 259aa86 is fine by me. Truthfully, I was following the example of pre-existing code in the toolbar without fully understanding the implications.

benmanley commented:
@paglias The only thing that needs doing pronto is the changes you were going to make to:

https://github.com/HabitRPG/habitrpg/blob/develop/src/controllers/groups.js#L257-L260

benmanley commented:
@wc8 @Mihakuu I agree with you both that it’s not ideal. 

I was hoping to be able to consolidate currency but I need the new icons to do that.

I don’t know what the consensus is on the subscription button, but that could be moved to the menu window if it doesn’t need to be front and centre.

Bailey could be moved to the notifications window for mobile. We could indicate that there was a new message via the notification icon and then you would see her peeking when you tapped on your notifications.

I think any two of these three things would do the trick.

lemoness commented:
Normally there would be a Bailey announcement - but it got lost among all
the WonderCon prep :/ Should I still announce it?

On Sat, Apr 19, 2014 at 1:11 PM, Ben Manley notifications@github.comwrote:

> @wc8 https://github.com/wc8 @Mihakuu https://github.com/Mihakuu I
> agree with you both that it’s not ideal.
> 
> I was hoping to be able to consolidate currency but I need the new icons
> to do that.
> 
> I don’t know what the consensus is on the subscription button, but that
> could be moved to the menu window if it doesn’t need to be front and centre.
> 
> Bailey could be moved to the notifications window for mobile. We could
> indicate that there was a new message via the notification icon and then
> you would see her peeking when you tapped on your notifications.
> 
> I think any two of these three things would do the trick.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/pull/3155#issuecomment-40879394
> .

ghost commented:
@benmanley in mobile browser notifications (guilds list) seem to have fixed max height. So if you have a device with let's say 1280x720 screen, the list occupies only some 60% of the notification window even if there are more of them - the list becomes scrollable, but only inside these 60% of space and the rest of the space above the "close" button remains empty.

benmanley commented:
Thanks @GitHubSphinx. I'll take a look at it.

benmanley commented:
@GitHubSphinx I’ve pushed a fix.

@wc8 @Mihakuu I’ve taken an executive decision and moved the subscribe button to the menu screen on mobile. I think the currency icons are still key to single line toolbar nirvana. I’ll look into that.

@paglias regarding 81ef260, I’ve pushed [a commit](https://github.com/HabitRPG/habitrpg/commit/254ac3d5101a5e64eb8bda3fd4ec7425491ac5e5) which consolidates those changes into the global styles I started and removes the need for the !important declarations.

@deilann Now that I’ve tidied up a few things, I’ll look at reinstating the collapse toolbar button.

I’m in two minds about tooltips for dropdowns that have headings on click. I know it’s better to label things, but they also get in the way a bit. I don’t know whether there’s already a convention I should be following.

ghost commented:
@benmanley should I already see the changes? It looks like everything is the same both on beta and on ordinary one http://i57.tinypic.com/2qdwzma.png

deilann commented:
I don't think it's been deployed yet.

On Thu, Apr 24, 2014 at 12:11 AM, Sphinx notifications@github.com wrote:

> @benmanley https://github.com/benmanley should I already see the
> changes? It looks like everything is the same both on beta and on ordinary
> one http://i57.tinypic.com/2qdwzma.png
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/pull/3155#issuecomment-41248854
> .

paglias commented:
@benmanley not directly related to this but settings page should have some padding

benmanley commented:
@paglias Looks like there were some typos in `settings.jade`. I've fixed them here: 7f984c859c79b1690416bc615042af186b98ec26

paglias commented:
@benmanley :+1: thanks

ghost commented:
@benmanley I know space is really an issue as far as the top bar is concerned, but could you please consider adding some character info to the section with gems, gold and silver?

The thing is that when you turn off the header (the character, stats and party bar) the only way to see you current health, for example, - is to open the character stats which may not be very convenient. So it'll be really great if when you turn off the header (only when it's turned off) the top bar will show you at least your health.

benmanley commented:
@GitHubSphinx I'll make a note and see what I can come up with.

deilann commented:
So, after getting over 1000 gold, I realized something kind of ridiculous.

Why is silver being shown? It seems kind of silly to only estimate the
amount of gold, but then show the exact number of silver.

I think it might be better to simply hide the silver once a player reaches
1000 gold. That would provide the space of like, 6 zeros.

On Fri, May 2, 2014 at 8:58 AM, Ben Manley notifications@github.com wrote:

> @GitHubSphinx https://github.com/GitHubSphinx I'll make a note and see
> what I can come up with.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/pull/3155#issuecomment-42047303
> .

benmanley commented:
True. It does seem a bit redundant at that point! (I’ve sometimes wondered if we need to see silver at all.) But there are a few arguments for keeping it:
1. Consistency – better that things don't pop in and out of the interface too often (other than Bailey). Could be a problem if you keep fluctuating between < and > 999.
2. Things tend to be more balanced/aesthetically pleasing in threes.
3. I’ve just posted an updated request for currency pixel art that will allow us to consolidate that bit of the bar. So, space wouldn’t be so much of a problem. 

What do you reckon?

BTW, I'm about to PR the toolbar hide/unhide button :-)

deilann commented:
I'd say that people who keep small amounts of gold around would like to see
their silver. If you have a lot of 1GP (<1 GP) rewards, your silver could
be a bigger deal. (Just like as a person who uses 100 GP rewards and keeps
around a bank of about 1000 GP, not quickly being able to see how close I
am to 1.1K is kind of annoying.)

I'd poll this actually.

And yay! Toolbar hide/unhide!

On Fri, May 2, 2014 at 11:35 AM, Ben Manley notifications@github.comwrote:

> True. It does seem a bit redundant at that point! (I’ve sometimes wondered
> if we need to see silver at all.) But there are a few arguments for keeping
> it:
> 1. Consistency – better that things don't pop in and out of the
>    interface too often (other than Bailey). Could be a problem if you keep
>    fluctuating between < and > 999.
> 2. Things tend to be more balanced/aesthetically pleasing in threes.
> 3. I’ve just posted an updated request for currency pixel art that
>    will allow us to consolidate that bit of the bar. So, space wouldn’t be so
>    much of a problem.
> 
> What do you reckon?
> 
> BTW, I'm about to PR the toolbar hide/unhide button :-)
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/pull/3155#issuecomment-42063511
> .

benmanley commented:
Yes. Didn't think of low value rewards.

If we stick with the 1.0K format, we could show the full gold amount on hover. Would that be a good solution to your issue?

deilann commented:
Yes, but it would still be weird to show the silver. Like, imagine if your
bank overview said that you had 1.1K dollars in your account and 38 cents.
You could easily open the statement and see how many dollars exactly, but
it would still be weird.

On Fri, May 2, 2014 at 1:15 PM, Ben Manley notifications@github.com wrote:

> Yes. Didn't think of low value rewards.
> 
> If we stick with the 1.0K format, we could show the full gold amount on
> hover. Would that be a good solution to your issue?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/pull/3155#issuecomment-42074432
> .

benmanley commented:
I know what you mean.

On Trello, Luciferian was suggesting an ingot for the gold icon. Perhaps we could use something like that for > 999 and ditch silver, like you say. The wider icon could cover the area vacated by silver. Wouldn’t save space, but might look more balanced. It could also act like an achievement badge of sorts.

My main priority is trying to wrangle the icons into a fairly consistent width.

deilann commented:
Seems legit!

On Sun, May 4, 2014 at 3:38 PM, Ben Manley notifications@github.com wrote:

> I know what you mean.
> 
> On Trello, Luciferian was suggesting an ingot for the gold icon. Perhaps
> we could use something like that for > 999 and ditch silver, like you say.
> The wider icon could cover the area vacated by silver. Wouldn’t save space,
> but might look more balanced. It could also act like an achievement badge
> of sorts.
> 
> My main priority is trying to wrangle the icons into a fairly consistent
> width.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/HabitRPG/habitrpg/pull/3155#issuecomment-42148372
> .

ghost commented:
@benmanley I just thought about my suggestion of having HP and XP somewhere on the bar... Actually it now seems more logic to me that a top bar might have a one-click "toggle character bar visibility" option under options or character menus. That will make a standard habitrpg theme more work-safe and usable at the same time. IMHO
